src/Stratis.Bitcoin.Tests/BlockPulling2/BlockPullerBehaviorTests.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using Moq;
+using Stratis.Bitcoin.BlockPulling2;
+using Stratis.Bitcoin.Configuration.Logging;
+using Stratis.Bitcoin.Interfaces;
+using Xunit;
+
+namespace Stratis.Bitcoin.Tests.BlockPulling2
+{
+    public class BlockPullerBehaviorTests
+    {
+        private readonly BlockPullerBehavior behavior;
+
+        public BlockPullerBehaviorTests()
+        {
+            var puller = new Mock<IBlockPuller>();
+
+            var ibdState = new Mock<IInitialBlockDownloadState>();
+            ibdState.Setup(x => x.IsInitialBlockDownload()).Returns(() => true);
+
+            var loggerFactory = new ExtendedLoggerFactory();
+            loggerFactory.AddConsoleWithFilters();
+
+            this.behavior = new BlockPullerBehavior(puller.Object, ibdState.Object, loggerFactory);
+        }
+
+        [Fact]
+        public void InitializesWithDefaultValues()
+        {
+            Assert.Equal(BlockPullerBehavior.SamplelessQualityScore, this.behavior.QualityScore);
+            Assert.Equal(0, this.behavior.SpeedBytesPerSecond);
+            Assert.Null(this.behavior.Tip);
+        }
+
+        [Fact]
+        public void QualityScoreCanGoToMinimum()
+        {
+            this.behavior.Penalize(10, 10);
+            this.behavior.RecalculateQualityScore(1000);
+
+            Assert.Equal(BlockPullerBehavior.MinQualityScore, this.behavior.QualityScore);
+        }
+
+        [Fact]
+        public void QualityScoreCanGoToMaximum()
+        {
+            this.behavior.AddSample(1000, 1);
+            this.behavior.RecalculateQualityScore(1);
+
+            Assert.Equal(BlockPullerBehavior.MaxQualityScore, this.behavior.QualityScore);
+        }
+
+        [Fact]
+        public void QualityScoreIsRelativeToBestSpeed()
+        {
+            this.behavior.AddSample(1000, 1);
+            this.behavior.RecalculateQualityScore(this.behavior.SpeedBytesPerSecond * 2);
+
+            Assert.True(this.DoubleEqual(0.5, this.behavior.QualityScore));
+        }
+
+        [Fact]
+        public void SpeedBytesPerSecondCalculatedCorrectly()
+        {
+            this.behavior.AddSample(1000, 1);
+            this.behavior.AddSample(2000, 1);
+            this.behavior.AddSample(1000, 10);
+            this.behavior.AddSample(500, 2);
+
+            // (500+1000+2000+1000) / (1+ 1 + 10 + 2) == 321.428
+            Assert.Equal(321, this.behavior.SpeedBytesPerSecond);
+        }
+
+        [Fact]
+        public void WhenIBDStateChangesMaxSamplesNuberIsRecalculated()
```

Comment:
Consistency

Also number

Reply:
Also should check whether the new average is correctly calculated

Reply:
>Also should check whether the new average is correctly calculated

That's covered in average calculator tests

Reply:
ok